pkg/operator/controllers/muo/muo_controller.go

```diff
@@ -0,0 +1,161 @@
+	pullSpec string = "rh.srep.muo.deploy.pullspec"
+)
+
+type Reconciler struct {
+	arocli        aroclient.Interface
+	kubernetescli kubernetes.Interface
+	dh            dynamichelper.Interface
+
+	readinessPollTime time.Duration
+	readinessTimeout  time.Duration
+}
+
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, kubernetescli kubernetes.Interface, dh dynamichelper.Interface) *Reconciler {
+	return &Reconciler{
+		arocli:        arocli,
+		kubernetescli: kubernetescli,
+		dh:            dh,
+
+		readinessPollTime: 10 * time.Second,
+		readinessTimeout:  5 * time.Minute,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.OperatorFlags.GetSimpleBoolean(enabled) {
+		// controller is disabled
+		condition := &operatorv1.OperatorCondition{
+			Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+			Status:  operatorv1.ConditionUnknown,
+			Message: "Controller is disabled",
+			Reason:  "ControllerDisabled",
+		}
+		err = conditions.SetCondition(ctx, r.arocli, condition, "master")
+		return reconcile.Result{}, err
+	}
+
+	managed := instance.Spec.OperatorFlags.GetWithDefault(managed, "")
+	deployer := NewMUODeployer(instance, r.kubernetescli, r.dh)
+
+	var condition *operatorv1.OperatorCondition
+
+	// If enabled and managed=true, install MUO
+	// If enabled and managed=false, remove the MUO deployment
+	// If enabled and managed is missing, do nothing
+	if strings.EqualFold(managed, "true") {
+		err = deployer.CreateOrUpdate(ctx)
+		if err != nil {
+			condition = &operatorv1.OperatorCondition{
+				Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+				Status:  operatorv1.ConditionFalse,
+				Message: err.Error(),
+				Reason:  "DeploymentFailed",
+			}
+			err = conditions.SetCondition(ctx, r.arocli, condition, "master")
+			return reconcile.Result{}, err
+		}
+
+		// Check that MUO has become ready, wait up to readinessTimeout (default 5min)
+		timeoutCtx, cancel := context.WithTimeout(ctx, r.readinessTimeout)
+		defer cancel()
+
+		err := wait.PollImmediateUntil(r.readinessPollTime, func() (bool, error) {
+			return deployer.IsReady(ctx)
+		}, timeoutCtx.Done())
+
+		if err != nil {
+			condition = &operatorv1.OperatorCondition{
+				Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+				Status:  operatorv1.ConditionFalse,
+				Message: "Deployment did not become ready",
+				Reason:  "DeploymentFailed",
+			}
+		} else {
+			condition = &operatorv1.OperatorCondition{
+				Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+				Status:  operatorv1.ConditionTrue,
+				Message: "Deployment is ready",
+				Reason:  "DeploymentSucceeded",
+			}
+		}
+	} else if strings.EqualFold(managed, "false") {
+		errs := deployer.Remove(ctx)
+		if len(errs) != 0 {
```

Comment:
See the comment above about getting rid of the conditions.

But if we decide that we actually want conditions - we should still requeue in case of errors on `deployer.Remove` to do a proper cleanup.